sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -259,8 +270,22 @@ object TypeCoercion {
     }
   }
 
-  private def haveSameType(exprs: Seq[Expression]): Boolean =
-    exprs.map(_.dataType).distinct.length == 1
+  private def haveSameType(exprs: Seq[Expression]): Boolean = {
```

Comment:
This is duplicated with `ComplexTypeMergingExpression.areInputTypesForMergingEqual`, can we unify them? We can
1. remove `hasSameType`. Any expression that needs to do this check should extend `ComplexTypeMergingExpression` and the TypeCoercion rule should call `areInputTypesForMergingEqual`.
2. remove `areInputTypesForMergingEqual`. `ComplexTypeMergingExpression` should only define the list of data types that need to be merged, and TypeCoercion rule should call `hasSameType(e.inputTypesForMerging)`